### PR TITLE
[ticket/13568] Validate imagick path as readable absolute path

### PR DESCRIPTION
**Pull request by marc1706**
Ticket: https://tracker.phpbb.com/browse/PHPBB3-13568

### CODE REVIEW

**Comment by nickvergessen** [comment]
Result is invalid for windows

**Comment by marc1706** [comment]
Added new validation option for absolute paths and also added functional tests for the imagick path setting.

File: phpBB/adm/index.php [hunk 1]
```diff
@@ -562,6 +562,42 @@ function validate_config_vars($config_vars, &$cfg_array, &$error)
 				}
 
 			break;
+
+			// Absolute file path
+			case 'wapath':
+			case 'apath':
```
Comment (nickvergessen) [inline phpBB/adm/index.php]: better useful names then short ones? :speak_no_evil:

File: phpBB/adm/index.php [hunk 2]
```diff
@@ -562,6 +562,42 @@ function validate_config_vars($config_vars, &$cfg_array, &$error)
...
+
+				$cfg_array[$config_name] = trim($cfg_array[$config_name]);
+
+				// Make sure no NUL byte is present...
+				if (strpos($cfg_array[$config_name], "\0") !== false || strpos($cfg_array[$config_name], '%00') !== false)
+				{
+					$cfg_array[$config_name] = '';
+					break;
+				}
+
```
Comment (nickvergessen) [inline phpBB/adm/index.php]: This is all duplicated from above.

I'd say try something like:

```php
$tmp =  ($config_definition['validate'] === 'wpath' || $config_definition['validate'] === 'path') ? $phpbb_root_path . $cfg_array[$config_name] : $cfg_array[$config_name];
```

and then use $tmp for validation

File: tests/functional/acp_attachments_test.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	 * @dataProvider data_imagick_path
+	 */
+	public function test_imagick_path($imagick_path, $expected)
+	{
+		$this->login();
+		$this->admin_login();
+
+		$crawler = self::request('GET', 'adm/index.php?i=attachments&mode=attach&sid=' . $this->sid);
+
+		$form = $crawler->selectButton('Submit')->form(array('config[img_imagick]'	=> $imagick_path));
```
Comment (nickvergessen) [inline tests/functional/acp_attachments_test.php]: While adding a test, please set all values to any extreme values :)

File: phpBB/adm/index.php (hunk 1 above)
Comment (marc1706) [inline phpBB/adm/index.php]: <nickvergessen> better use apath
<nickvergessen> or absolute path

But sure, I'll use more descriptive ones

File: tests/functional/acp_attachments_test.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+		$crawler = self::submit($form);
+		$this->assertContains($expected, $crawler->text());
```
Comment (nickvergessen) [inline tests/functional/acp_attachments_test.php]: please use a filter

File: tests/functional/acp_attachments_test.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * @group functional
+ */
+class phpbb_functional_acp_attachments_test extends phpbb_functional_test_case
+{
+	public function data_imagick_path()
+	{
+		return array(
+			array('/usr/bin', 'Configuration updated successfully'),
+			array('/usr/bin/', 'Configuration updated successfully'),
+			array('C:\Windows\system32', 'The entered path “C:\Windows\system32” does not exist.'),
```
Comment (nickvergessen) [inline tests/functional/acp_attachments_test.php]: Test fails on windows :/

File: tests/functional/acp_attachments_test.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * @group functional
+ */
+class phpbb_functional_acp_attachments_test extends phpbb_functional_test_case
+{
+	public function data_imagick_path()
+	{
+		return array(
+			array('/usr/bin', 'Configuration updated successfully'),
```
Comment (nickvergessen) [inline tests/functional/acp_attachments_test.php]: test fails on windows

**Comment by nickvergessen** [comment]
you need to create 2 tests and skip them if on windows/not on windows
(or 3 because of mac?)

**Comment by marc1706** [comment]
I'd say skip these tests if not on linux.

File: tests/functional/acp_attachments_test.php [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+			array('/usr/bin/', 'Configuration updated successfully'),
+			array('C:\Windows\system32', 'The entered path “C:\Windows\system32” does not exist.'),
+			array('/usr/nope', 'The entered path “/usr/nope” does not exist.'),
+			array('mkdir /usr/test', 'The entered path “mkdir /usr/test” does not exist.'),
+			array('/usr/bin/which', 'The entered path “/usr/bin/which” is not a directory.'),
```
Comment (cs278) [inline tests/functional/acp_attachments_test.php]: What about a test case for `.` which should fail, but doesn't look like it will.

**Comment by nickvergessen** [comment]
> I'd say skip these tests if not on linux.

yes and then add another test suite that works on windows.
We want to make sure this works on IIS right?!

**Comment by marc1706** [comment]
Test case and unit tests that cover more paths will be added for a develop-ascraeus patch

**Comment by marc1706** [comment]
Tests are passing on windows for me now.

**Comment by marc1706** [comment]
@nickvergessen @bantu if this PR looks fine to you I'll create one based on this for ascraeus that also adds unit tests

**Comment by nickvergessen** [comment]
yes please

**Comment by marc1706** [comment]
@bantu PHP 5.2 builds seem to not work again

**Comment by bantu** [comment]
@marc1706 I have been told 5.2 is permanently gone and 5.3.3 will follow.

**Comment by nickvergessen** [comment]
http://docs.travis-ci.com/user/build-environment-updates/2015-02-03/#PHP-VM

> ### PHP VM
* PHP updates: 5.4.35 → 5.4.37, 5.5.19 → 5.5.21, 5.6.3 → 5.6.5
* HHVM 3.4.0 → 3.5.0
* xdebug 2.2.5 → 2.2.7
* PHP 5.2.17 is removed

**Comment by Nicofuma** [comment]
http://docs.travis-ci.com/user/build-environment-updates/2015-02-03/

**Comment by Nicofuma** [comment]
merged in b9db47e

**Pull request closed without merging**